tests/Qdrant.Client.Tests/FilteringTests.cs

```diff
@@ -0,0 +1,310 @@
+		var expected = new Filter
+		{
+			MustNot =
+			{
+				new Condition { HasId = new HasIdCondition { HasId = { new PointId { Num = 1 } } } },
+				new Condition
+				{
+					Field = new FieldCondition { Key = "foo", Match = new Match { Keyword = "hello" } }
+				}
+			}
+		};
+
+		Assert.Equal(expected, actual);
+	}
+
+	[Fact]
+	public void LogicalOr_conditions_produces_condition_filter_with_two_should_conditions()
+	{
+		var actual = HasId(1) || MatchKeyword("foo", "hello");
+		var expected = new Filter
+		{
+			Should =
+			{
+				new Condition { HasId = new HasIdCondition { HasId = { new PointId { Num = 1 } } } },
+				new Condition
+				{
+					Field = new FieldCondition { Key = "foo", Match = new Match { Keyword = "hello" } }
+				},
+			}
+		};
+
+		Assert.Equal(expected, actual);
+	}
+
+	[Fact]
+	public void LogicalOr_conditions_with_logicalNot_produces_condition_filter_must_not_condition()
+	{
+		var actual = !(HasId(1) || MatchKeyword("foo", "hello"));
+		var expected = new Filter
+		{
+			MustNot =
+			{
+				new Condition
+				{
+					Filter = new Filter
+					{
+						Should =
+						{
+							new Condition { HasId = new HasIdCondition { HasId = { new PointId { Num = 1 } } } },
+							new Condition
+							{
+								Field = new FieldCondition { Key = "foo", Match = new Match { Keyword = "hello" } }
+							}
+						}
+					}
+				}
+			}
+		};
+
+		Assert.Equal(expected, actual);
+	}
+
+	[Fact]
+	public void LogicalOr_logicalNot_conditions_produces_condition_filter_must_not_condition()
+	{
+		var actual = !HasId(1) || !MatchKeyword("foo", "hello");
+		var expected = new Filter
+		{
+			Should =
+			{
+				new Condition
+				{
+					Filter = new Filter
+					{
+						MustNot =
+						{
+							new Condition { HasId = new HasIdCondition { HasId = { new PointId { Num = 1 } } } }
+						}
+					}
+				},
+				new Condition
+				{
+					Filter = new Filter
+					{
+						MustNot =
+						{
+							new Condition
+							{
+								Field = new FieldCondition { Key = "foo", Match = new Match { Keyword = "hello" } }
+							}
+						}
+					}
+				}
+			}
+		};
+
+		Assert.Equal(expected, actual);
+	}
+
+	[Fact]
+	public void LogicalOr_with_condition_or_logicalNot_condition_produces_condition_filter_with_should_conditions()
+	{
+		var actual = HasId(1) || !MatchKeyword("foo", "hello");
+		var expected = new Filter
+		{
+			Should =
+			{
+				new Condition { HasId = new HasIdCondition { HasId = { new PointId { Num = 1 } } } },
+				new Condition
+				{
+					Filter = new Filter
+					{
+						MustNot =
+						{
+							new Condition
+							{
+								Field = new FieldCondition { Key = "foo", Match = new Match { Keyword = "hello" } }
+							}
+						}
+					}
+				},
+			}
+		};
+
+		Assert.Equal(expected, actual);
+	}
+
+	[Fact]
+	public void Complex_condition()
+	{
+		var actual = ((HasId(1) || HasId(2)) && !MatchKeyword("foo", "hello")) ||
+					 (HasId(3) && (MatchKeyword("foo", "hello") || !IsNull("foo")));
+
+		var expected =
```

Comment:
Lovely...  This really should have been a small text-based DSL rather than an object representation of boolean logic.

Definitely drives home the value of this PR.